src/fides/api/ops/service/connectors/saas_connector.py

```diff
@@ -427,56 +433,75 @@ def run_consent_request(
         policy: Policy,
         privacy_request: PrivacyRequest,
         identity_data: Dict[str, Any],
-        consent_preferences: List[Consent],
+        consent_preferences: List[PrivacyRequestConsentPreference],
     ) -> bool:
         """Execute a consent request. Return whether the consent request to the third party succeeded.
-        Return True if 200 OK
+        Should only propagate either the entire set of opt in or opt out requests.
+        Return True if 200 OK. Raises a SkippingConsentPropagation exception if no action is taken
+        against the service.
         """
         logger.info(
             "Starting consent request for node: '{}'",
             node.address.value,
         )
         self.set_privacy_request_state(privacy_request, node)
         query_config = self.query_config(node)
+        should_opt_in: Optional[bool] = should_opt_in_to_service(
+            self.configuration.system, consent_preferences
+        )
 
-        for consent_options in consent_preferences:
-            query_config.action = (
-                "opt_in" if consent_options.opt_in else "opt_out"
-            )  # For logging purposes
+        if should_opt_in is None:
+            logger.info(
+                "Skipping consent requests on node {}: No actionable consent preferences defined",
+                node.address.value,
+            )
+            raise SkippingConsentPropagation(
+                f"Skipping consent propagation for node {node.address.value} - no actionable consent preferences defined"
+            )
 
-            consent_requests: List[
-                SaaSRequest
-            ] = self._get_consent_requests_by_preference(consent_options.opt_in)
+        matching_consent_requests: List[
+            SaaSRequest
+        ] = self._get_consent_requests_by_preference(should_opt_in)
 
-            if not consent_requests:
-                logger.info(
-                    "Skipping consent requests on node {}: No '{}' requests defined",
-                    node.address.value,
-                    query_config.action,
-                )
-                continue
+        query_config.action = (
+            "opt_in" if should_opt_in else "opt_out"
+        )  # For logging purposes
 
-            for consent_request in consent_requests:
-                self.set_saas_request_state(consent_request)
+        if not matching_consent_requests:
+            logger.info(
+                "Skipping consent requests on node {}: No '{}' requests defined",
+                node.address.value,
+                query_config.action,
+            )
+            raise SkippingConsentPropagation(
```

Comment:
The main change for the saas connectors besides the propagation logic is that I am adding *skipped* logs instead of *completed* logs if we don't actually send a request to the service, to distinguish between what services were affected.  This is for reporting purposes coming up.